flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBKeyedStateBackend.java

```diff
@@ -1991,43 +1999,71 @@ public int numStateEntries() {
 		return count;
 	}
 
+	/**
+	 * This class is not thread safety.
+	 */
 	private static class RocksIteratorWrapper<K> implements Iterator<K> {
 		private final RocksIterator iterator;
 		private final String state;
 		private final TypeSerializer<K> keySerializer;
 		private final int keyGroupPrefixBytes;
+		private final byte[] namespaceBytes;
+		private K nextKey;
 
 		public RocksIteratorWrapper(
 				RocksIterator iterator,
 				String state,
 				TypeSerializer<K> keySerializer,
-				int keyGroupPrefixBytes) {
+				int keyGroupPrefixBytes,
+				byte[] namespaceBytes) {
 			this.iterator = Preconditions.checkNotNull(iterator);
 			this.state = Preconditions.checkNotNull(state);
 			this.keySerializer = Preconditions.checkNotNull(keySerializer);
 			this.keyGroupPrefixBytes = Preconditions.checkNotNull(keyGroupPrefixBytes);
+			this.namespaceBytes = Preconditions.checkNotNull(namespaceBytes);
+			this.nextKey = null;
 		}
 
 		@Override
 		public boolean hasNext() {
-			return iterator.isValid();
+			final int namespaceBytesLength = namespaceBytes.length;
+			while (nextKey == null && iterator.isValid()) {
+				try {
+					boolean namespaceValid = true;
+					byte[] key = iterator.key();
+					if (key.length >= namespaceBytesLength + keyGroupPrefixBytes) {
```

Comment:
We could decompose this loop by introducing a helper method `isMatchingNameSpace(byte[] key, byte[] nameSpaceBytes)`. I think that makes the control flow easier to read.

Reply:
addressed.